### DISCUSSION TITLE
How to create reusable controller components with TS, in real world?

### DISCUSSION THREAD

**Discussion by patjakubik** [react-hook-form/react-hook-form#6878] [Q&A] [answered]
Hey everyone,
I want to create a reusable controller for each form component from my UI Kit, I started with a text field and have few issues with it, which [documentation doesn't really cover](https://react-hook-form.com/api/usecontroller). 

<img width="604" alt="CleanShot 2021-10-23 at 11 24 43@2x" src="https://user-images.githubusercontent.com/37963339/138551210-13f23141-55e3-486c-822b-532ad368c4c3.png">
<img width="639" alt="CleanShot 2021-10-23 at 11 26 08@2x" src="https://user-images.githubusercontent.com/37963339/138551213-2f8334bc-23ed-4155-8d44-d21ae6a72cdb.png">

1. How can I pass props from ControlledTextField component to the TextField component, such as label, disabled, placeholder etc.?

2. How can I make it truly reusable if UseControllerProps requires passing form fields type?

CodeSandbox: https://codesandbox.io/s/quirky-star-j7kzs?file=/src/App.tsx

I'm using [Shopify Polaris UI Kit](https://github.com/Shopify/polaris-react/), but it doesn't seem to be relevant.

**Comment by bluebill1049** [comment]
try this next beta version: https://github.com/react-hook-form/react-hook-form/releases/tag/v7.18.0-next.0

**Reply by patjakubik** [reply]
Thanks for your reply, although I'm not sure how to use this. Does that address issue 1 or 2? I still can see form type being passed to the controller, so not sure how this helps.

**Reply by patjakubik** [reply]
Just added CodeSandbox to the main post, hope that illustrates my issue better.

**Comment by lundn** [comment, marked as answer]
1. How can I pass props from ControlledTextField component to the TextField component, such as label, disabled, placeholder etc.?
You're not just talking about basic react props here, like `<ControlledTextField label={"Label"} ... />` and then passing it to `TextField` through `props.label` inside of `ControlledTextField`?

2. How can I make it truly reusable if UseControllerProps requires passing form fields type?
Does this mean that you want a component that can integrate with any form library, without being tightly coupled with said library? Much like what `TextField` from Shopify Polaris UI Kit is?

**Reply by patjakubik** [reply]
Hey @lundn , thanks for taking your time to reply! 

1. Yes, exactly. I'm getting TypeScript errors here.

2. No, I mean that currently my `ControlledTextField` takes `props` argument, which type is `UseControllerProps<Settings>` - the `Settings` part is a type of my Settings form inputs.

The question here is: How can I reuse `ControlledTextField` component across multiple forms in my app, if `UseControllerProps` requires passing type specific to a particular form? Does that make sense?

**Reply by lundn** [reply]
1. Ofc `label:string` would need to be added to some interface/type that ControlledTextField uses. I guess you would either create a new type e.g `InputFieldProps` (which contains label,placeholder and such) and use it in the component, like `props: InputFieldProps & UseControllerProps<T>`

2. I think what you're after is Generics. Like `function ControlledTextField<T>(props: UseControllerProps<T>)`.

**Reply by patjakubik** [reply]
Got it! Thank you very much!
